core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeComponent(user: Identity, action: WhiskActionMetaData, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) {
+
+    val exec = action.toExecutableWhiskAction
+    if (action.annotations.get("conductor").isDefined && exec.isDefined) {
+      // composition
+      invokeComposition(user, exec.head, payload, None, Some(session.activationId), Some(session)) // non-blocking
+      // invokeComposition will take care of accounting and continuations
+    } else {
+      session.accounting.components += 1
+      exec
+        .map { exec =>
+          // not a sequence, not a composition
+          invokeSimpleAction(
+            user,
+            exec,
+            payload,
+            Some(action.limits.timeout.duration + 1.minute),
+            Some(session.activationId))
+        }
+        .getOrElse {
+          // sequence
+          val SequenceExecMetaData(components) = action.exec
+          invokeSequence(user, action, components, payload, None, Some(session.activationId), false, 0).map(r => r._1)
+        }
+        .onComplete {
+          case Failure(t) =>
+            val response = ActivationResponse.whiskError(compositionActivationFailure)
+            completeAppActivation(user, session, response)
+          case Success(Left(activationId)) =>
+            session.logs += activationId
+            val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+            completeAppActivation(user, session, response)
+          case Success(Right(activation)) =>
+            session.logs += activation.activationId
+            session.duration += activation.duration.getOrElse(
+              activation.end.toEpochMilli - activation.start.toEpochMilli)
+            activation.annotations.get("limits") map { limitsAnnotation =>
+              limitsAnnotation.asJsObject.getFields("memory") match {
+                case Seq(JsNumber(memory)) =>
+                  session.maxMemory = Math.max(session.maxMemory, memory.toInt)
+              }
+            }
+
+            // invoke continuation
+            invokeConductor(user, Some(activation.resultAsJson), session)
+        }
+    }
+  }
+
+  /**
+   * Creates an activation for a composition and writes it back to the datastore.
+   * Completes the associated promise if any.
+   * Resumes caller if any.
+   */
+  private def completeAppActivation(user: Identity, session: Session, response: ActivationResponse)(
+    implicit transid: TransactionId): Unit = {
+
+    // compute max memory
+    val sequenceLimits = Parameters(
+      WhiskActivation.limitsAnnotation,
+      ActionLimits(session.action.limits.timeout, MemoryLimit(session.maxMemory MB), session.action.limits.logs).toJson)
+
+    // set causedBy if not topmost
+    val causedBy = if (session.cause.isDefined) {
```

Comment:
this is just a map on session.cause.map(...) simplifies the if/else some/none.

Reply:
OK